File: brain-js/workspaces/brain/src/node.ts
import type { ZodType } from "zod";

import type { InvocationError } from "./error.ts";
import { Access, Internet } from "./runtime.ts";

export interface NodeOptions<I, O> {
  readonly name: string;
  readonly input: ZodType<I>;
  readonly output: ZodType<O>;
  readonly prompt: string;
  readonly access?: Access;
  readonly internet?: boolean;
}

export interface NodeSpec<I, O> {
  readonly name: string;
  readonly prompt: string;
  readonly access: Access;
  readonly internet: Internet;
  readonly inputSchema: ZodType<I>;
  readonly outputSchema: ZodType<O>;
}

/** A named agent operation with schema-validated input and output. */
export class Node<I, O> {
  /** @internal */
  readonly _spec: NodeSpec<I, O>;

  /** @internal */
  constructor(options: NodeOptions<I, O>) {
    this._spec = Object.freeze({
      name: options.name,
      prompt: options.prompt,
      access: options.access ?? Access.ReadOnly,
      internet: options.internet === true ? Internet.Enabled : Internet.Disabled,
      inputSchema: options.input,
      outputSchema: options.output,
    });
  }

  get name(): string {
    return this._spec.name;
  }

  /** Creates a request for this node without executing it. */
  withInput(input: I): NodeInvocation<I, O> {
    return new NodeInvocation(this, input);
  }
}

/** Creates a typed agent node with read-only, offline defaults. */
export function node<I, O>(options: NodeOptions<I, O>): Node<I, O> {
  return new Node(options);
}

/** A typed request to invoke a particular node. */
export class NodeInvocation<I, O> {
  readonly node: Node<I, O>;
  readonly input: I;

  /** @internal */
  constructor(node: Node<I, O>, input: I) {
    this.node = node;
    this.input = input;
  }
}

export type NodeOutcome<I, O> =
  | { readonly ok: true; readonly value: O }
  | {
      readonly ok: false;
      readonly error: InvocationError;
      readonly input: I;
      readonly invocation: number;
    };

File: brain/src/node.rs
use std::{marker::PhantomData, sync::atomic::{AtomicU64, Ordering}};

use crate::{Access, Internet, InvocationError};

/// Creates a typed agent node with read-only, offline defaults.
///
/// The name must be static so the resulting node remains [`Copy`].
pub fn node<I, O>(name: &'static str) -> Node<I, O> {
    Node::new(name)
}

static NEXT_NODE_ID: AtomicU64 = AtomicU64::new(1);

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) struct NodeSpec {
    pub id: u64,
    pub name: &'static str,
    pub prompt: &'static str,
    pub access: Access,
    pub internet: Internet,
}

/// A named agent operation with typed input and output.
#[derive(Debug)]
pub struct Node<I, O> {
    pub(crate) spec: NodeSpec,
    marker: PhantomData<fn(I) -> O>,
}

impl<I, O> Copy for Node<I, O> {}

impl<I, O> Clone for Node<I, O> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<I, O> Node<I, O> {
    pub fn new(name: &'static str) -> Self {
        Self {
            spec: NodeSpec {
                id: NEXT_NODE_ID.fetch_add(1, Ordering::Relaxed),
                name,
                prompt: "",
                access: Access::default(),
                internet: Internet::default(),
            },
            marker: PhantomData,
        }
    }

    /// Sets the static prompt used for every invocation of this node.
    pub fn prompt(mut self, prompt: &'static str) -> Self {
        self.spec.prompt = prompt;
        self
    }

    pub fn access(mut self, access: Access) -> Self {
        self.spec.access = access;
        self
    }

    pub fn internet(mut self, internet: Internet) -> Self {
        self.spec.internet = internet;
        self
    }

    pub fn name(&self) -> &str {
        self.spec.name
    }

    /// Creates one invocation without consuming the reusable node handle.
    pub fn with(&self, input: I) -> NodeInvocation<I, O> {
        NodeInvocation { node: *self, input }
    }
}

/// A typed request to invoke a particular node.
#[derive(Debug)]
pub struct NodeInvocation<I, O> {
    pub(crate) node: Node<I, O>,
    pub(crate) input: I,
}

/// A failed node invocation that retains ownership of its original input.
#[derive(Debug)]
pub struct NodeFailure<I> {
    input: I,
    error: InvocationError,
    invocation: usize,
}

impl<I> NodeFailure<I> {
    pub(crate) fn new(input: I, error: InvocationError, invocation: usize) -> Self {
        Self {
            input,
            error,
            invocation,
        }
    }

    pub fn input(&self) -> &I {
        &self.input
    }

    pub fn error(&self) -> &InvocationError {
        &self.error
    }

    pub const fn invocation(&self) -> usize {
        self.invocation
    }

    pub fn into_input(self) -> I {
        self.input
    }

    pub fn into_error(self) -> InvocationError {
        self.error
    }

    pub fn into_parts(self) -> (I, InvocationError) {
        (self.input, self.error)
    }
}
